appinventor/components/src/com/google/appinventor/components/runtime/BarChartDataModel.java

```diff
@@ -0,0 +1,187 @@
+      }
+    } catch (Exception e) {
+      // 2-tuples are invalid when null entries are present, or if
+      // the number of entries is not sufficient to form a pair.
+      // TODO: Show toast error notification
+    }
+
+    return null;
+  }
+
+  @Override
+  protected void setDefaultStylingProperties() {
+
+  }
+
+  @Override
+  public void removeEntry(int index) {
+    // Entry exists; remove it
+    if (index >= 0) {
+      // If the index of the Entry to remove is the last Entry,
+      // we can simply remove the Entry from the Data Series' values.
+      if (index == entries.size() - 1) {
+        entries.remove(index);
+      } else {
+        // If the Entry to be removed is not the last Entry, we
+        // have to instead set the Y value of the Entry to be
+        // removed to 0 in order to preserve the Bar Chart Data
+        // properties (x values sorted and correspond to indexes)
+        entries.get(index).setY(0f);
+      }
+    }
+  }
+
+  @Override
+  public void addTimeEntry(YailList tuple) {
+    // TODO: Currently, this implementation breaks the sorted x value
+    // TODO: property (the START X VALUE is no longer 0). A potential
+    // TODO: fix could be shifting the values instead of removing them,
+    // TODO: zeroing out values on removal (which would result in too
+    // TODO: many entries, however) or by handling the case where
+    // TODO: data adding is added after/during Real Time Data import.
+
+    // If the entry count of the Data Series entries exceeds
+    // the maximum allowed time entries, then remove the first one
+    if (entries.size() >= maximumTimeEntries) {
+      entries.remove(0);
+    }
+
+    entries.add((BarEntry)getEntryFromTuple(tuple));
+  }
+
+  @Override
+  protected boolean areEntriesEqual(Entry e1, Entry e2) {
+    // To avoid (unlikely) cast exceptions, check that
+    // both the entries are of instance BarEntry, and return
+    // false if that is not the case.
+    if (!(e1 instanceof BarEntry && e2 instanceof BarEntry)) {
+      return false;
+    }
+
+    // Due to Bar grouping affecting the x values of the
+    // entries, we need to floor the entries to get rid
+    // of the decimal part. The offset is guaranteed to be
+    // less than 1 (based on the source code and the chosen
+    // granularity and spacing)
+    return Math.floor(e1.getX()) == Math.floor(e2.getX()) &&
+        e1.getY() == e2.getY();
+  }
+
+  @Override
+  public YailList getTupleFromEntry(Entry entry) {
+    // Create a list with the X and Y values of the entry, and
+    // convert the generic List to a YailList. Since Bar Chart
+    // grouping adds offsets to the x value (which are expected
+    // to be below 1), the x value needs to be floored.
+    List tupleEntries = Arrays.asList((float)Math.floor(entry.getX()), entry.getY());
+    return YailList.makeList(tupleEntries);
+  }
+
+  // TODO: FindEntryIndex can be optimized by making use of the property
```

Comment:
@ewpatton You are right, the comment is a bit out of place. This is basically referring to the fact that the FindEntryIndex method in the superclass is doing a linear search, while it can be optimized in the BarChartDataModel to use the index directly in O(1) time instead, since it has this property.
Instead of leaving this out here, I can satisfy the TODO, or just remove this altogether, or maybe add it to the Javadoc.